File: LoginRadiusSDK/LoginRadiusField.m
//
//  LoginRadiusField.m
//  Pods
//
//  Created by Thompson Sanjoto on 2017-06-02.
//
//

#import <Foundation/Foundation.h>
#import "LoginRadiusField.h"

NSString *const kProviderEndpoint = @"Endpoint";
NSString *const kProviderName = @"Name";
@implementation LoginRadiusField

- (instancetype)init:(NSDictionary*)dictionary
{
    self = [super init];
    if (self) {
        _type = [self setTypeWithString:[dictionary objectForKey:@"type"]];
        _option = (_type == OPTION) ? [self initializeOptions:[dictionary objectForKey:@"options"]] : nil;
        _name = [dictionary objectForKey:@"name"];
        _display = [dictionary objectForKey:@"display"];
        _permission = [((NSString *)[dictionary objectForKey:@"permission"]) isEqualToString:@"w"]? WRITE : READ;
        _rules = ([dictionary objectForKey:@"rules"]) ? [self initializeRules:[dictionary objectForKey:@"rules"]] : nil;
        }
    
    return self;
}

-(instancetype)initWithSocialSchema:(NSDictionary *)dictionary
{
    self = [super init];
    if(![dictionary[kProviderEndpoint] isKindOfClass:[NSNull class]]){
        self.endpoint = dictionary[kProviderEndpoint];
    }
    if(![dictionary[kProviderName] isKindOfClass:[NSNull class]]){
        self.providerName = dictionary[kProviderName];
    }
    return self;
}

+ (NSArray<NSString *> *_Nonnull) addressFields {
    return @[@"address1",@"address2",@"city",@"country",@"postalcode",@"region",@"state"];
}


- (LoginRadiusFieldType) setTypeWithString:(NSString*)fieldString
{
    LoginRadiusFieldType fieldType = HIDDEN;
    if ([fieldString isEqualToString:@"string"]) {
        fieldType = STRING;
    } else if ([fieldString isEqualToString:@"option"]) {
        fieldType = OPTION;
    } else if ([fieldString isEqualToString:@"multi"]) {
        fieldType = MULTI;
    } else if ([fieldString isEqualToString:@"password"]) {
        fieldType = PASSWORD;
    } else if ([fieldString isEqualToString:@"hidden"]) {
        fieldType = HIDDEN;
    } else if ([fieldString isEqualToString:@"email"]) {
        fieldType = EMAILID;
    } else if ([fieldString isEqualToString:@"text"]) {
        fieldType = TEXT;
    }
    return fieldType;
}

- (NSDictionary*) initializeOptions:(NSArray*) options
{
    NSMutableDictionary *newOptions = [[NSMutableDictionary alloc] init];
    
    for(NSDictionary *option in options)
    {
        [newOptions setObject:option[@"text"] forKey:option[@"value"]];
    }

    return [newOptions copy];
}

-(NSArray<LoginRadiusFieldRule *> *)initializeRules: (NSString*) fullRuleStr
{
    //if nil or empty string
    if (!fullRuleStr || [fullRuleStr length] <= 0)
    {
        return nil;
    }

    NSMutableArray<LoginRadiusFieldRule *> *rules = [[NSMutableArray<LoginRadiusFieldRule *> alloc] init];
    
    //this is to avoid custom validation regex that containts "|"
    NSString *customValidation = nil;
    if ([fullRuleStr containsString:@"custom_validation"])
    {

        NSRange start = [fullRuleStr rangeOfString:@"custom_validation["];
        NSUInteger startIndex = start.location+start.length;
        int openSquares = 0;
        //iteratate through the content of custom_validation[.... until it founds the ] pair
        while (openSquares >= 0 || startIndex > [fullRuleStr length] )
        {
            if ([[fullRuleStr substringWithRange:NSMakeRange(startIndex,1)] isEqualToString:@"["])
            {
                openSquares++;
            }else if([[fullRuleStr substringWithRange:NSMakeRange(startIndex,1)] isEqualToString:@"]"])
            {
                openSquares--;
            }
        
            startIndex++;
        }
        
        NSAssert((openSquares==-1), @"Invalid Custom Regex");
        
        //save "custom_validation[...]"
        customValidation = [fullRuleStr substringWithRange:NSMakeRange(start.location,startIndex-start.location)];
        //delete it from the fullRuleStr with empty string
        fullRuleStr = [fullRuleStr stringByReplacingCharactersInRange:NSMakeRange(start.location,startIndex-start.location) withString:@""];
    }
    
    NSMutableArray<NSString *> *rulesStr = [[fullRuleStr componentsSeparatedByString:@"|"] mutableCopy];
    
    if (customValidation)
    {
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"length > 0"];
        rulesStr = [[rulesStr filteredArrayUsingPredicate:predicate] mutableCopy]; //delete any element with empty array
        [rulesStr addObject:customValidation]; //add custom_validation[...] into rules array
    }
    
    for (NSString *ruleStr in rulesStr)
    {
        LoginRadiusFieldRule *newRule = [[LoginRadiusFieldRule alloc] init:ruleStr];
        if ([newRule type] == valid_date)
        {
            _type = DATE;
        }
        [rules addObject:[[LoginRadiusFieldRule alloc] init:ruleStr]];
    }
    
    return [rules copy];
}

-(NSString *)description
{
    NSString *lrfieldstr = [NSString stringWithFormat:@"<LoginRadiusField, type: %@, name: %@, display: %@, required:%@",LoginRadiusFieldType([self type]), [self name], [self display], self.isRequired ? @"YES":@"NO"];
    
    if ([self option])
    {
        lrfieldstr = [lrfieldstr stringByAppendingString:[NSString stringWithFormat:@", option:%@",[self option]]];
    }
    
    lrfieldstr = [lrfieldstr stringByAppendingString:@">"];

    return lrfieldstr;
}

-(BOOL) getIsRequired
{
    BOOL req = NO;
    
    for (LoginRadiusFieldRule *rule in _rules)
    {
        if ([rule type] == required)
        {
            req = YES;
            break;
        }
    }
    
    return req;
}

-(NSString*) typeToString
{
    return LoginRadiusFieldType([self type]);
}

@end

File: Sources/LoginRadiusField.swift

//
//  LoginRadiusField.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation


public enum LoginRadiusFieldType {
    case string
    case option
    case multi
    case password
    case hidden
    case email
    case text
    case date
}

public enum LoginRadiusFieldPermission {
    case read
    case write
}

public class LoginRadiusField {
    public var type: LoginRadiusFieldType = .hidden
    public var name: String
    public var display: String
    public var option: [String: String]?
    public var permission: LoginRadiusFieldPermission
    public var rules: [LoginRadiusFieldRule]?
    public var isRequired: Bool {
        return rules?.contains { $0.type == .required } ?? false
    }
    public var endpoint: String?
    public  var providerName: String?
    
    public init(dictionary: [String: Any]) {
        self.type = .hidden
        self.name = ""
        self.display = ""
        self.permission = .read
        
        if let typeString = dictionary["type"] as? String {
            self.type = setTypeWithString(fieldString: typeString)
        }
        
        if let name = dictionary["name"] as? String {
            self.name = name
        }
        
        if let display = dictionary["display"] as? String {
            self.display = display
        }
        
        if let permissionString = dictionary["permission"] as? String {
            self.permission = (permissionString == "w") ? .write : .read
        }
        
        if let options = dictionary["options"] as? [[String: Any]], self.type == .option {
            self.option = initializeOptions(options: options)
        }
        if let rulesString = dictionary["rules"] as? String {
            self.rules = initializeRules(fullRuleStr: rulesString)
        }
    }
    
    public init(socialSchema: [String: Any]) {
        self.type = .hidden
        self.name = ""
        self.display = ""
        self.permission = .read
        
        if let endpoint = socialSchema["Endpoint"] as? String {
            self.endpoint = endpoint
        }
        
        if let providerName = socialSchema["Name"] as? String {
            self.providerName = providerName
        }
    }
    
    public static func addressFields() -> [String] {
        return ["address1", "address2", "city", "country", "postalcode", "region", "state"]
    }
    
    func setTypeWithString(fieldString: String) -> LoginRadiusFieldType {
        switch fieldString {
        case "string": return .string
        case "option": return .option
        case "multi": return .multi
        case "password": return .password
        case "hidden": return .hidden
        case "email": return .email
        case "text": return .text
        case "date": return .date
        default: return .hidden
        }
    }
    
    public func initializeOptions(options: [[String: Any]]) -> [String: String] {
        var newOptions = [String: String]()
        for option in options {
            if let value = option["value"] as? String, let text = option["text"] as? String {
                newOptions[value] = text
            }
        }
        return newOptions
    }
    
    func initializeRules(fullRuleStr: String) -> [LoginRadiusFieldRule]? {
        // If nil or empty string
        if fullRuleStr.isEmpty {
            return nil
        }
        
        var rules = [LoginRadiusFieldRule]()
        var mutableFullRuleStr = fullRuleStr // Create a mutable copy of the input string
        
        // This is to avoid custom validation regex that contains "|"
        var customValidation: String? = nil
        
        if mutableFullRuleStr.contains("custom_validation") {
            if let start = mutableFullRuleStr.range(of: "custom_validation[") {
                let startIndex = start.upperBound
                var openSquares = 0
                
                // Iterate through the content of custom_validation[.... until it finds the ] pair
                var currentIndex = startIndex
                while openSquares >= 0 && currentIndex < mutableFullRuleStr.endIndex {
                    let currentChar = mutableFullRuleStr[currentIndex]
                    if currentChar == "[" {
                        openSquares += 1
                    } else if currentChar == "]" {
                        openSquares -= 1
                    }
                    
                    currentIndex = mutableFullRuleStr.index(after: currentIndex)
                }
                
                assert(openSquares == -1, "Invalid Custom Regex")
                customValidation = String(mutableFullRuleStr[start.lowerBound..<currentIndex])
                
                // Delete it from the fullRuleStr with an empty string
                mutableFullRuleStr.replaceSubrange(start.lowerBound..<currentIndex, with: "")
            }
        }
        
        var rulesStr = mutableFullRuleStr.components(separatedBy: "|")
        
        if let customValidation = customValidation {
            rulesStr = rulesStr.filter { !$0.isEmpty }
            rulesStr.append(customValidation)
        }
        
        for ruleStr in rulesStr {
            let newRule = LoginRadiusFieldRule(ruleStr: ruleStr)
            if newRule.type == .valid_date {
                type = .date // This line appears to be problematic; please review it
            }
            rules.append(LoginRadiusFieldRule(ruleStr: ruleStr))
        }
        
        return rules
    }
    
    
    public  var description: String {
        var lrFieldStr = "<LoginRadiusField, type: \(typeToString), name: \(name), display: \(display), required: \(isRequired ? "YES" : "NO")"
        if let option = option {
            lrFieldStr += ", option: \(option)"
        }
        lrFieldStr += ">"
        return lrFieldStr
    }
    
    public func typeToString() -> String {
        switch type {
        case .string: return "string"
        case .option: return "option"
        case .multi: return "multi"
        case .password: return "password"
        case .hidden: return "hidden"
        case .email: return "email"
        case .text: return "text"
        case .date: return "date"
        }
    }
}
